builtins/web/fetch/request-response.cpp

```diff
@@ -2684,6 +2727,68 @@ JSObject *Response::create_incoming(JSContext *cx, host_api::HttpIncomingRespons
   return self;
 }
 
+ResponseFutureTask::ResponseFutureTask(const HandleObject request, host_api::FutureHttpIncomingResponse *future)
+      : request_(request), future_(future) {
+    auto res = future->subscribe();
+
+    MOZ_ASSERT(!res.is_err(), "Subscribing to a future should never fail");
+    handle_ = res.unwrap();
+}
+
+bool ResponseFutureTask::run(api::Engine *engine) {
+  // MOZ_ASSERT(ready());
+  JSContext *cx = engine->cx();
+
+  const RootedObject request(cx, request_);
+  RootedObject response_promise(cx, Request::response_promise(request));
+
+  auto res = future_->maybe_response();
+  if (res.is_err()) {
+    api::throw_error(cx, FetchErrors::FetchNetworkError);
+    return RejectPromiseWithPendingError(cx, response_promise);
+  }
+
+  auto maybe_response = res.unwrap();
+  if(!maybe_response.has_value()) {
+    return cancel(engine);
+  }
+
+  auto response = maybe_response.value();
+  RootedObject response_obj(cx, Response::create_incoming(cx, response));
+  if (!response_obj) {
+    return false;
+  }
+
+  RequestOrResponse::set_url(response_obj, RequestOrResponse::url(request));
+  RootedValue response_val(cx, ObjectValue(*response_obj));
+  if (!ResolvePromise(cx, response_promise, response_val)) {
+    return false;
+  }
+
+  return cancel(engine);
+}
+
+bool ResponseFutureTask::cancel(api::Engine *engine) {
+  // TODO(TS): implement
+  handle_ = -1;
+  return true;
+}
+
+bool ResponseFutureTask::abort(api::Engine *engine) {
```

Comment:
I think we should be aborting the response object as well if it exists, so that the code like this work:

```js
    const controller = new AbortController();
    const signal = controller.signal;

    const log = [];
    const response = await fetch('../resources/data.json', { signal });

    controller.abort();

    const bodyPromise = response[bodyMethod]();

    await Promise.all([
      bodyPromise.catch(() => log.push(`${bodyMethod}-reject`)),
      Promise.resolve().then(() => log.push('next-microtask'))
    ]);

    await promise_rejects_dom(t, "AbortError", bodyPromise);
```

I don't think I have a good instinct where that should happen. It's really hard to map this part of implementation to the spec. Our fetch does not follow the spec very closely.

Reply:
Oh right: the spec doesn't ever assume lazy creation of objects like this. I now remember having that same problem with the implementation of marking bodies as used. Can you maybe integrate these two things? I.e., reuse `Slots::BodyUsed` for this by changing it to an enum of `Usable, Used, Disturbed`, instead of introducing a new separate flag?

Reply:
Ah, true :( In that case let's leave things as they are. I have some ideas on how we could save a slot here, but we'd have a lot of lower-hanging fruit than this one :)